hie-plugin-api/Haskell/Ide/Engine/Ghc.hs

```diff
@@ -130,131 +129,185 @@ srcErrToDiag df rfm se = do
+loadFile rfm t = do
+    withProgress "loading" NotCancellable $ \f -> (captureDiagnostics rfm $ BIOS.loadFileWithMessage (Just $ toMessager f) t)
+
+-- | Actually load the module if it's not in the cache
+setTypecheckedModule_load :: Uri -> IdeGhcM (IdeResult (Diagnostics, AdditionalErrs))
+setTypecheckedModule_load uri =
   pluginGetFile "setTypecheckedModule: " uri $ \fp -> do
-    fileMap <- GM.getMMappedFiles
-    debugm $ "setTypecheckedModule: file mapping state is: " ++ show fileMap
-    rfm <- GM.mkRevRedirMapFunc
-    let
-      ghcErrRes msg = ((Diagnostics Map.empty, [T.pack msg]),Nothing,Nothing)
-      progTitle = "Typechecking " <> T.pack (takeFileName fp)
     debugm "setTypecheckedModule: before ghc-mod"
+    debugm "Loading file"
+    mapped_fp <- persistVirtualFile uri
+    liftIO $ copyHsBoot fp mapped_fp
+    rfm <- reverseFileMap
     -- TODO:AZ: loading this one module may/should trigger loads of any
     -- other modules which currently have a VFS entry.  Need to make
     -- sure that their diagnostics are reported, and their module
     -- cache entries are updated.
     -- TODO: Are there any hooks we can use to report back on the progress?
-    ((Diagnostics diags', errs), mtm, mpm) <- withIndefiniteProgress progTitle NotCancellable $ GM.gcatches
-      (GM.getModulesGhc' (myWrapper rfm) fp)
-      (errorHandlers ghcErrRes (return . ghcErrRes . show))
-    debugm "setTypecheckedModule: after ghc-mod"
-
+    (Diagnostics diags', errs, mmods) <- loadFile rfm (fp, mapped_fp)
+    debugm "File, loaded"
     canonUri <- toNormalizedUri <$> canonicalizeUri uri
     let diags = Map.insertWith Set.union canonUri Set.empty diags'
-    diags2 <- case (mpm,mtm) of
-      (Just pm, Nothing) -> do
-        debugm $ "setTypecheckedModule: Did get parsed module for: " ++ show fp
-        cacheModule fp (Left pm)
-        debugm "setTypecheckedModule: done"
-        return diags
-
-      (_, Just tm) -> do
+    debugm "setTypecheckedModule: after ghc-mod"
+    debugm ("Diags: " <> show diags')
+    let collapse Nothing = (Nothing, [])
+        collapse (Just (n, xs)) = (n, xs)
+
+    case collapse mmods of
+      --Just (Just pm, Nothing) -> do
+      --  debugm $ "setTypecheckedModule: Did get parsed module for: " ++ show fp
+       -- cacheModule fp (Left pm)
+       -- debugm "setTypecheckedModule: done"
+      --  return diags
+
+      (Just _tm, ts) -> do
         debugm $ "setTypecheckedModule: Did get typechecked module for: " ++ show fp
-        sess <- fmap GM.gmgsSession . GM.gmGhcSession <$> GM.gmsGet
+        --sess <- fmap GM.gmgsSession . GM.gmGhcSession <$> GM.gmsGet
 
         -- set the session before we cache the module, so that deferred
         -- responses triggered by cacheModule can access it
-        modifyMTS (\s -> s {ghcSession = sess})
-        cacheModule fp (Right tm)
+
+        Session sess <- GhcT pure
+        modifyMTS (\s -> s {ghcSession = Just sess})
```

Comment:
I think we agreed that this `modify` can be deleted as it was only used by completions.

Reply:
How should completions work without this?

Reply:
I thought completion information came from the `GlobalRdEnv` in `TypecheckedModule`?

Reply:
We still need it to get the type when we resolve the completion.

It is also used for haddock.